### DISCUSSION TITLE
Match status with precinct by precinct

### DISCUSSION THREAD

**Discussion by kuriwaki** [kuriwaki/cvr_harvard-mit_scripts#138] [General]
[Note: This is not really an issue to "fix" per se, but I am using this as an importantFYI].

@JeffreyBLewis has matched about the precincts of 240 counties to VEST shapefiles. 

The results are currently in Sheet3 of our Google Sheet, linked [here](https://docs.google.com/spreadsheets/d/1PnrWgZP0aQ9MoyhvRfH1p3KLhZLWBKCiPWNa8Y0j9JQ/edit#gid=1691714058) for now. 

Basically Jeff has gone state by state, first string-matching the precinct names from the CVRs to the VEST precinct names, then if that fails, trying to match them as good as possible by the Biden/Trump vote count in that precinct (this is my attempt at summarizing his hundreds of lines of code). He used the MEDSL precinct returns as a second check.

He has a database of precinct matches, but what I'm showing here is a county-level summary. See the following table:

|colour | counties with precinct-level exact match | all  counties |
|:------|-----------------:|---:|
|green  |               132| 143|
|yellow |                 7|  27|
|red    |                22|  69|

The second column indicates counties where all non-missing precincts' vote totals matched to VEST. This says that among 143 green counties (i.e. the totals exactly match), 132 **also** were cases in which all the non-missing precinct pairs matched exactly in their Biden/Trump vote with VEST. There are 7 + 22 = 29 such precinct-level successes in the yellow and red counties too: I think these are cases where we are missing entire precincts, but for each non-missing the totals line up. 

So, 132 /143 is great. We should try to figure out what the remaining 11 are.

The states included in this tabulation for now are "CALIFORNIA", "COLORADO", "GEORGIA", "ILLINOIS", "MICHIGAN", 
"NEVADA", "OHIO", "UTAH".

Some code that created the linked table's variables:

```
     summarize(n_precincts_cvr = n_distinct(precinct.cvr),
                n_precincts_vest = n_distinct(precinct.vest),
                n_precincts_medsl = n_distinct(precinct.medsl) *
                                      if_else(!all(is.na(precinct.medsl)), 1, NA),
                n_precincts_sov = n_distinct(precinct.medsl) *
                                    if_else(!all(is.na(precinct.sov)), 1, NA),
                mean_vote_dist = mean(vote_dist, na.rm = TRUE),
                max_vote_dist = max(vote_dist, na.rm = TRUE),
                n_zero_vote_dist = sum(vote_dist == 0, na.rm = TRUE),
                across(mean_vote_dist:n_zero_vote_dist, 
                       \(x) if_else(n_precincts_vest == 0, NA, x)),
                max_medsl_vs_vest_diff = max(mv_dist, na.rm = TRUE),
                .by = c("county")) |> 
                arrange(county)
```

**Comment by JeffreyBLewis** [comment]
I added counties from AK, FL, ND, and NJ for which precinct info is currently available to the spreadsheet (95 in total).  NJ is particularly challenging.  I didn't fill in the county status because I am not sure what the status of each of those counties is at the moment.

**Comment by JeffreyBLewis** [comment]
I went ahead and used the colors from sheet2 to fill in the count status for the new sheet 3 entries.

**Comment by JeffreyBLewis** [comment]
Shiro's description of the process at the top here more or less captures it.   I would add that the merging by names is a fuzzy merge and that the matching does not sequentially fall back on vote totals, but uses the vote totals in combination with the names in establishing all matches.   Also, for counties where the merge is tricky, there is a lot of preprocessing of the precinct names (for example where the CVR precinct labels are precinct portions that must be aggregated to match the precinct-level data).   So, this has been a state and county-specific effort that involves some standard record linkage tricks, but also a lot of visual inspection of results from counties in which the names and votes do not perfectly across the CVR and VEST to make code adjustments and to ensure match quality which is why, as Shiro says, there is a lot of (not very tidy) code.

**Comment by kuriwaki** [comment]
I update the google sheet by hand, but the underlying table is now created in this repo, [01b_precinct-merge-results.R](https://github.com/kuriwaki/cvr_harvard-mit_scripts/blob/ad2ae0161642f00cda3d1dcad2683d1a8178c731/R/combine/01b_precinct-merge-results.R) (initial commit, to be merged with main). 

I pull `to-parquet/JBL/precinct_merge_summary.csv` and merge it with the parquet results. @JeffreyBLewis , I noticed you put in your new states like NJ to `precinct_merge_summary_additional.csv` . can you simply combine that into the original precinct_merge_summary.csv? That would help simplify my pipeline

**Comment by JeffreyBLewis** [comment]
I think I was only doing it that way to avoid having to rerun the whole thing when adding new counties/states. That whole setup could be streamlined.  The issue is that there is a lot of bespoke state-by-state and county-by-county stuff going on so that code is not terribly slick and mostly just kludges its way to a solution on a case-by-case basis.  I wonder if what we want to do is have the building of the VEST-to-CVR crosswalk as a separate (sub)project and only use the assembled crosswalk as part of the ongoing production process.

**Comment by kuriwaki** [comment]
> "I wonder if what we want to do is have the building of the VEST-to-CVR crosswalk as a separate (sub)project and only use the assembled crosswalk as part of the ongoing production process."

Yes, ideally we would merge that cross-walk upstream into the metadata (so each CVR has a `precinct` and a `precinct_for_vest` variable). That does seem ideal, but it does seem like a lot of consistent work, and lower relative priority.

I've merged three counties in Wisconsin and that is a state that is hard. In Dane county VEST reports votes for specific wards (Town of Albion 001, Town of Albion 002), but the county and the CVR reports votes for groups of wards (Town of Albion Wards 001-002). I wonder how VEST managed to unbundle the two.
